File: backend/tests/helpers/auth.helper.ts
/**
 * Authentication Helper for Integration Tests
 *
 * Provides functions to obtain JWT tokens from Keycloak for test users.
 */

import axios from 'axios';
import { TEST_CONFIG } from '../setup';

export interface TestUser {
  id: string;
  email: string;
  password: string;
  realm: string;
  role: string;
}

// Token cache to avoid repeated Keycloak requests
const tokenCache = new Map<string, { token: string; expiresAt: number }>();

/**
 * Gets an access token for a test user from Keycloak.
 * Uses caching to avoid redundant token requests.
 */
export async function getAccessToken(user: TestUser): Promise<string> {
  const cacheKey = `${user.realm}:${user.email}`;
  const cached = tokenCache.get(cacheKey);

  // Return cached token if still valid (with 30 second buffer)
  if (cached && cached.expiresAt > Date.now() + 30000) {
    return cached.token;
  }

  const tokenUrl = `${TEST_CONFIG.keycloakUrl}/realms/${user.realm}/protocol/openid-connect/token`;

  try {
    const response = await axios.post(
      tokenUrl,
      new URLSearchParams({
        grant_type: 'password',
        client_id: TEST_CONFIG.clientId,
        username: user.email,
        password: user.password,
      }),
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      }
    );

    const token = response.data.access_token;
    const expiresIn = response.data.expires_in || 300; // Default 5 minutes

    // Cache the token
    tokenCache.set(cacheKey, {
      token,
      expiresAt: Date.now() + expiresIn * 1000,
    });

    return token;
  } catch (error: any) {
    const message = error.response?.data?.error_description || error.message;
    throw new Error(`Failed to get access token for ${user.email}: ${message}`);
  }
}

/**
 * Gets an access token for a specific tenant and role.
 */
export async function getTokenForRole(
  tenant: 'alpha' | 'beta',
  role: 'admin' | 'manager' | 'member'
): Promise<string> {
  const user = TEST_CONFIG.users[tenant][role];
  return getAccessToken(user);
}

/**
 * Creates an Authorization header value for a test user.
 */
export async function getAuthHeader(user: TestUser): Promise<string> {
  const token = await getAccessToken(user);
  return `Bearer ${token}`;
}

/**
 * Clears the token cache. Useful between test runs.
 */
export function clearTokenCache(): void {
  tokenCache.clear();
}

/**
 * Generates an invalid/expired token for negative testing.
 */
export function getInvalidToken(): string {
  // This is a malformed JWT that will fail verification
  return 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMn0.invalid_signature';
}

/**
 * Generates an expired token structure for testing.
 * Note: This won't be valid since it's not signed by Keycloak.
 */
export function getExpiredToken(): string {
  // A token that claims to be expired (exp in the past)
  const header = Buffer.from(JSON.stringify({ alg: 'RS256', typ: 'JWT' })).toString('base64url');
  const payload = Buffer.from(
    JSON.stringify({
      sub: 'test-user',
      iss: `${TEST_CONFIG.keycloakUrl}/realms/alpha`,
      exp: Math.floor(Date.now() / 1000) - 3600, // 1 hour ago
      iat: Math.floor(Date.now() / 1000) - 7200, // 2 hours ago
    })
  ).toString('base64url');
  const signature = 'fake_signature';

  return `${header}.${payload}.${signature}`;
}

File: admin-service/tests/helpers/auth_helper.py
"""
Authentication helper for admin service tests.

Provides functions to obtain tokens and authenticate with Keycloak.
"""

import os
import httpx
from typing import Optional

KEYCLOAK_URL = os.environ.get("KEYCLOAK_URL", "http://localhost:8081")
KEYCLOAK_ADMIN_USER = os.environ.get("KEYCLOAK_ADMIN_USER", "admin")
KEYCLOAK_ADMIN_PASSWORD = os.environ.get("KEYCLOAK_ADMIN_PASSWORD", "admin")

# Token cache
_admin_token_cache: Optional[str] = None


async def get_keycloak_admin_token() -> str:
    """
    Gets an admin token from Keycloak for administrative operations.
    Uses caching to avoid repeated requests.
    """
    global _admin_token_cache

    if _admin_token_cache:
        return _admin_token_cache

    token_url = f"{KEYCLOAK_URL}/realms/master/protocol/openid-connect/token"

    async with httpx.AsyncClient() as client:
        response = await client.post(
            token_url,
            data={
                "grant_type": "password",
                "client_id": "admin-cli",
                "username": KEYCLOAK_ADMIN_USER,
                "password": KEYCLOAK_ADMIN_PASSWORD,
            },
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )

        if response.status_code != 200:
            raise Exception(f"Failed to get admin token: {response.text}")

        _admin_token_cache = response.json()["access_token"]
        return _admin_token_cache


async def get_user_token(realm: str, username: str, password: str) -> str:
    """
    Gets a user token from Keycloak for a specific realm.
    """
    token_url = f"{KEYCLOAK_URL}/realms/{realm}/protocol/openid-connect/token"

    async with httpx.AsyncClient() as client:
        response = await client.post(
            token_url,
            data={
                "grant_type": "password",
                "client_id": "taskhub-app",
                "username": username,
                "password": password,
            },
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )

        if response.status_code != 200:
            raise Exception(f"Failed to get user token: {response.text}")

        return response.json()["access_token"]


def clear_token_cache():
    """Clears the token cache."""
    global _admin_token_cache
    _admin_token_cache = None


async def wait_for_keycloak(max_attempts: int = 30, delay_seconds: int = 2) -> bool:
    """
    Waits for Keycloak to be ready.
    """
    import asyncio

    health_url = f"{KEYCLOAK_URL}/health/ready"

    async with httpx.AsyncClient() as client:
        for i in range(max_attempts):
            try:
                response = await client.get(health_url)
                if response.status_code == 200:
                    print(f"Keycloak ready after {i + 1} attempts")
                    return True
            except Exception:
                pass
            await asyncio.sleep(delay_seconds)

    return False
